### ISSUE TITLE
How to ignore hidden object deserialization by using derived macros?

### ISSUE THREAD

**Issue by LasmGratel** [rakaly/jomini#103]
I got `Error(InvalidSyntax { msg: "hidden object must start with a key", offset: 3560 })` while trying to deserialize a Stellaris mod technology definition, file contains
```
tech_xxx {
weight_modifier {
...
    any_owned_planet = {
					OR = {
						has_building = building_mote_harvesters
						building_mote_harvesting_traps_2 // <-- ERROR HERE
...
```
But I didn't define `weight_modifier` field in my technology struct

So my questions are: 
1. why should I deal with a hidden object while I simply want to ignore that field? If so, how can I deal with it?
2. TextTape could be an alternative but it is too complicated to use, especially when dealing with a multi layer nested object. Any real examples using TextTape to deserialize a game file?
3. Could the parser allow to be continued after a syntax error to be found, and return a None instead of an error? Might not be clear but it's like HTML parsers vs XHTML parsers.

**Comment by nickbabcock** [comment]
To answer questions directly;

>  1. why should I deal with a hidden object while I simply want to ignore that field? If so, how can I deal with it?

Unfortunately the mixing of object field and array values inside a single block is noted [unsupported syntax](https://github.com/rakaly/jomini#unsupported-syntax), and is some of the only known syntax to cause a failure.

It is possible to workaround it by preprocessing the input (eg: removing offending line or turning or some other mechanism for transforming the block into recognized syntax). But this is not a savory thought.

> 2. TextTape could be an alternative but it is too complicated to use, especially when dealing with a multi layer nested object. Any real examples using TextTape to deserialize a game file?

Yeah, there here's [an example](https://github.com/pdx-tools/pdx-tools/blob/81d877c0bc6aca4b7f41c04b81677ddc07c8d61d/src/cli/src/cmd/compile_assets/religion.rs#L31-L56) of parsing eu4 religion file with text tape. But it does require parsing to succeed.
 
> 3. Could the parser allow to be continued after a syntax error to be found, and return a None instead of an error? Might not be clear but it's like HTML parsers vs XHTML parsers.

[Rust allows you to convert an error to None](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok)

---

So what's the solution? Obviously this shouldn't be thought of as invalid syntax if the game engine accepts it. And I agree.

Previously I thought this syntax was extremely rare, but if mods are using it, maybe it is not so uncommon.

And I've been reluctant to support the syntax as it may require performance or ergonomic degradation (it may be hard to figure out the most intuitive container to deserialize a mixed object/array block)

Perhaps there is a convenient compromise. I have a couple ideas rolling around that may be worth a shot, but this is a hard problem to solve.

**Comment by LasmGratel** [comment]
Thank you, in the mean time I will try to make a workaround.

2022年9月27日 上午10:14:50 Nick Babcock ***@***.***>:

> To answer questions directly;
> 
1. > why should I deal with a hidden object while I simply want to ignore that field? If so, how can I deal with it?
> 
> Unfortunately the mixing of object field and array values inside a single block is noted unsupported syntax[https://github.com/rakaly/jomini#unsupported-syntax], and is some of the only known syntax to cause a failure.
> 
> It is possible to workaround it by preprocessing the input (eg: removing offending line or turning or some other mechanism for transforming the block into recognized syntax). But this is not a savory thought.
> 
2. > TextTape could be an alternative but it is too complicated to use, especially when dealing with a multi layer nested object. Any real examples using TextTape to deserialize a game file?
> 
> Yeah, there here's an example[https://github.com/pdx-tools/pdx-tools/blob/81d877c0bc6aca4b7f41c04b81677ddc07c8d61d/src/cli/src/cmd/compile_assets/religion.rs#L31-L56] of parsing eu4 religion file with text tape. But it does require parsing to succeed.
> 
3. > Could the parser allow to be continued after a syntax error to be found, and return a None instead of an error? Might not be clear but it's like HTML parsers vs XHTML parsers.

**Comment by nickbabcock** [comment]
So I have a fix in the works (#104). It's not ready yet. But feel free to upload the problem file so I can use it as a test case.
